### ISSUE TITLE
Why are 'maybe' bounds not permitted in trait objects?

### ISSUE THREAD

**Issue by alexreg** [rust-lang/rust#59656]
For example,

```rust
type Foo = dyn Send + ?Sized;
```

gives the error `` `?Trait` is not permitted in trait object types ``.

Is there any good reason for this rule?

I ask mainly because I'm try to enforce this for trait objects that use trait aliases in the compiler, but this error is currently raised during AST validation, and would be very difficult to do during typeck, since information about 'maybe' bounds is only included in the HIR and not the type system.

CC @Centril

**Comment by Centril** [comment]
Also relevant:
+ duplicates in trait objects, e.g. should `dyn Copy + Copy` be allowed?
+ order dependence in trait objects wrt. the "principal trait", e.g. should `dyn Send + Copy` be allowed?

**Comment by alexreg** [comment]
@Centril Yes, though w.r.t. the 2nd point, I think the compiler team already agreed that removing the concept of principal trait is a good thing. It was really only ever an implementation detail I believe... though if the lang team can ratify this, all the better.

As for the first, I'm personally less concerned about changing that behaviour right now (it doesn't really simplify the implementation), but the original point of this issue is definitely something we want to discuss. :-)

**Comment by withoutboats** [comment]
As far as I know, there's no good reason for the restriction except that it avoids having to make sure that `dyn Send` and `dyn Send + ?Sized` are treated as the same type.

In my opinion, it would be fine - if someone wanted to make the effort - to allow all of these sort of meaningless variations, as long as we unify them to the same type as the equivalent bound:

* Multiple of the same trait
* All orderings of the same set of bounds
* All marker traits except for `Sized`
* Adding `?Sized`, which means nothing

Its important to note that `Sized` behaves differently from other marker traits: adding `+ Sized` should not be permitted (because this type will not meet that bound) and adding `+ ?Sized` means nothing (because, unlike normal bounds, the dyn trait bounds are all assumed to be `?Sized` because they must allow it).

If someone wants to put in the effort to implement this properly, it'd be nice to move forward with allowing all of this. What's trickier to allow is multiple bounds which are not all marker traits, because we have to figure out the runtime representation of that.

cc @nikomatsakis, to see if he agrees with this

**Comment by alexreg** [comment]
Thanks @withoutboats. That makes sense to me too... although the fact that `dyn Sized` is allowed seems to suggest a counterexample, which has rather befuddled @Centril and me as to exactly what the semantics are. But I agree, let's get Niko's input on this in any case.

**Comment by nikomatsakis** [comment]
I think I basically agree with everything @withoutboats wrote except that I'm not sure what they mean by "marker trait" precisely. Moreover, I think that "maybe bounds" is really kind of a... misnomer. A `?Sized` bound is not a *bound* at all -- it's suppressing a default `Sized` bound. Since we don't have that default in the context of trait objects, it makes no sense to suppress it (particularly since it would be an error).

I'm not aware of `?` being usable with any other traits -- or am I forgetting something?

I would say that if we *did* later add a default (let's call it `?QBit`, to use @pnkfelix's example of yore, although perhaps I'm misremembering), and we said that `dyn Foo` implied automatically `dyn (Foo + QBit)`, then it would make sense to have `dyn Foo + ?QBit` to suppress that default.

**Comment by kryptan** [comment]
Allowing `?Sized` in trait objects would be confusing because you can't actually cast unsized values to trait objects. If I see `&dyn Trait` I assume that I can cast `&T` to it if `T` implements `Trait`. If I saw `&(dyn Send + ?Sized)` I would therefore think that I can cast `&str` to it but it is not allowed. In my mind model, traits listed after `dyn` refer both to traits that underling object implements and to traits that `dyn ...` value itself implements. `?Sized` would be special because it only means that `dyn ...` type itself is unsized but underlying value still cannot be unsized.

**Comment by alexreg** [comment]
What are the current thoughts on this, generally?
